Allow fee estimation to work with lower fees

ajtowns opened the pull request:
Changes the minimum bucket for fee estimation from 1000 satoshi per kilobyte to 4 satoshi per kilobyte, and upgrades any saved fee estimation data.

Empact commented on src/policy/fees.cpp:
```diff
@@ -195,6 +202,45 @@ TxConfirmStats::TxConfirmStats(const std::vector<double>& defaultBuckets,
+    : buckets(defaultBuckets), bucketMap(defaultBucketMap)
+{
+    decay = txstatsSource.decay;
+    scale = txstatsSource.scale;
```
nit: these could go in the initializer list

Empact commented on src/policy/fees.cpp:
```diff
@@ -195,6 +202,45 @@ TxConfirmStats::TxConfirmStats(const std::vector<double>& defaultBuckets,
+    avg.resize(buckets.size());
+
+    unsigned int src = 0;
+    for (unsigned int dst = 0; dst < buckets.size(); dst++) {
```
nit: `++dst` here and elsewhere

Empact commented on src/policy/fees.cpp:
```diff
@@ -195,6 +202,45 @@ TxConfirmStats::TxConfirmStats(const std::vector<double>& defaultBuckets,
+        }
+
+        src++;
+        if (src >= txstatsSource.buckets.size()) break;
```
nit: can move these conditions into the `for`, so that it manages both `src` and `dst`.

ajtowns commented:
The really high fee rates were due to not handling the buckets/bucketMap objects correctly, believe that's fixed now.

ajtowns commented:
Dropped the WIP tag, this should be good to review/nitpick now.

I couldn't come up with a reasonable way to add functional tests for this change; if someone has a good idea, suggestions/patches welcome :) Though... I guess just generating an estimates file directly in python, pointing bitcoind at it, and seeing what estimates bitcoind comes up with after upgrading it without ever actually putting txes in the blockchain could work okay...

ajtowns committed:
Add CBlockPolicyEstimator::CheckConsistent helper

ajtowns committed:
Make fee buckets go much lower

ajtowns committed:
Add comment to CBlockPolicyEstimator::Read

ajtowns committed:
[refactor] Add CBlockPolicyEstimator::InitBucketMap helper

ajtowns committed:
Allow extending fee estimation buckets

practicalswift commented on src/policy/fees.cpp:
```diff
@@ -195,6 +202,56 @@ TxConfirmStats::TxConfirmStats(const std::vector<double>& defaultBuckets,
+    : buckets(defaultBuckets), bucketMap(defaultBucketMap),
+      decay(txstatsSource.decay), scale(txstatsSource.scale)
+{
+    const unsigned int maxPeriods = txstatsSource.confAvg.size();
```
Nit: Make `maxPeriods` – and the three following `i` – `size_t` instead to avoid implicit conversions?

ajtowns commented:
Reworked based on a suggestion from @Sjors; could still use some functional tests to check that loading a particular set of fee rates produces an expected set of fee estimates.

practicalswift commented on src/policy/fees.h:
```diff
@@ -168,25 +168,27 @@ class CBlockPolicyEstimator
 public:
     /** Create new BlockPolicyEstimator and initialize stats tracking classes with default values */
-    CBlockPolicyEstimator();
+    CBlockPolicyEstimator(double min_bucket_feerate = MIN_BUCKET_FEERATE);
```
Should be `explicit`?

ajtowns replied:
Yep.

ajtowns committed:
[tests] Check estimations when loaded from file

Makes feature_fee_estimates load a fixed fee_estimates.dat and
checks that the estimates given afterwards are also fixed.

stevenroose commented:
Is this work abandoned?

Empact commented:
Concept ACK

ajtowns commented:
Had stalled this on writing some tests. Having now written a test, it's found a bug, so now trying to fix that... :)

ajtowns commented:
Rebased, test case added as first commit, should be reviewable.

Note that this only makes estimations work for fee rates below 1sat/byte if you ever actually accept fee rates that low into your mempool. To make your mempool accept/relay transactions that cheap requires setting `-minrelaytxfee` or changing `DEFAULT_MIN_RELAY_TX_FEE`; and to have your wallet use fees that low means setting`-mintxfee` or changing `DEFAULT_TRANSACTION_MINFEE` and potentially also `WALLET_INCREMENTAL_RELAY_FEE`. Doing those things now for mainnet would probably just be a spam vector, rather than particularly useful IMHO.

MarcoFalke commented:
> Doing those things now for mainnet would probably just be a spam vector

Also open a huge surface for privacy leaks

stevenroose commented on src/policy/fees.h:
```diff
@@ -166,6 +166,7 @@ class CBlockPolicyEstimator
      * The MIN_BUCKET_FEERATE and MAX_BUCKET_FEERATE should just be set to
      * the lowest and highest reasonable feerate we might ever want to track.
      */
+    static constexpr double MIN_BUCKET_FEERATE = 4;
```
The PR OP says "8 satoshi". I don't know if this conflicts with the concept ACKs on the PR.

ajtowns replied:
Right, if I remember correctly, the original patchset didn't preserve the estimate data, but did have a lower bound of exactly 8sat/kvb. The patchset now does preserve the old estimates, but as a result only approximates the lower bound -- choose 4 gives an error of 0.8% whereas 8 gave an error of 4.7%. Updated the PR description.

stevenroose commented on src/policy/fees.cpp:
```diff
@@ -599,7 +599,8 @@ CBlockPolicyEstimator::CBlockPolicyEstimator(double min_bucket_feerate)
     : nBestSeenHeight(0), firstRecordedHeight(0), historicalFirst(0), historicalBest(0), trackedTxs(0), untrackedTxs(0)
 {
     assert(0 < min_bucket_feerate);
-    static_assert(BASE_BUCKET_FEERATE > 0, "Base feerate must be nonzero");
```
Will `BASE_BUCKET_FEERATE` still be used?

ajtowns replied:
BASE_BUCKET_FEERATE defines the bucket alignment, MIN and MAX then define the extremities.

stevenroose commented on src/policy/fees.h:
```diff
@@ -260,6 +261,9 @@ class CBlockPolicyEstimator
     int CheckConsistent() const EXCLUSIVE_LOCKS_REQUIRED(m_cs_fee_estimator);
 
+    /** Initialise buckets up to base */
+    size_t ExtendBucketsUpTo(double base, double min) EXCLUSIVE_LOCKS_REQUIRED(m_cs_fee_estimator);
```
I would appreciate the documentation on this method to be a little bit more descriptive. IIUC it adds buckets on the lower end from min up to base?

ajtowns replied:
Update the comment

stevenroose commented on src/policy/fees.h:
```diff
@@ -163,25 +163,26 @@ class CBlockPolicyEstimator
+     * The MIN_BUCKET_FEERATE and MAX_BUCKET_FEERATE should just be set to
+     * the lowest and highest reasonable feerate we might ever want to track.
      */
-    static constexpr double MIN_BUCKET_FEERATE = 1000;
```
Hmm, I think something weird happened in the separation of the commits. Here you seem to remove the `MIN_BUCKET_FEERATE` constant, while in https://github.com/bitcoin/bitcoin/pull/13990/commits/c68f796c41b9b145c9c1b2c24aea505362ad1b63 you're adding it again.

ajtowns replied:
Prior to this PR, `MIN_BUCKET_FEERATE` defines both the lower bound of the buckets, and how they're aligned (that is, all buckets are at `FEE_SPACING**n * MIN_BUCKET_FEERATE`). This commit renames `MIN_BUCKET_FEERATE` to `BASE_BUCKET_FEERATE` because the alignment will continue to be based on it, even as the minimum gets reduced in the later commit that reintroduce `MIN_BUCKET_FEERATE` at a new, lower value.

stevenroose reviewed (commented):
I felt obliged to take a look at the entire PR as it seemed that I prompted renewed interest. I'm not fully familiar with the bucket system, but all seemed reasonable.

prayank23 commented:
> Doing those things now for mainnet would probably just be a spam vector, rather than particularly useful IMHO.

@ajtowns what if those things only work if mempool has transactions less than 10 MB and switch to default it greater than 10 MB

> Also open a huge surface for privacy leaks

@MarcoFalke what kind of privacy leaks are possible with such changes?

prayank23 commented:
Got no response for the last comment. Still trying to understand what can go wrong if fee rate below 1 sat/vByte is allowed

Related SE questions: https://bitcoin.stackexchange.com/questions/99089/can-we-use-fee-rate-less-than-1-sat-vbyte

https://bitcoin.stackexchange.com/questions/99975/what-bitcoin-core-options-can-be-changed-without-a-restart

Related tweet: https://twitter.com/pedr0fr/status/1326716647449702401

If miners revenue is a concern or incentives in Bitcoin, maybe we will have to look at lot of charts, consider everything and discuss with more people especially who use bitcoin like coinjoin etc. I don't see spam and privacy as reasons or maybe need more details.

rebroad commented:
@ajtowns sorry for the noob-like question, but why is so much code needed to simply reduce the minfee...? I'd have thought it ought to be a one-line change for this. What am I missing?

GeneFerneau commented on src/policy/fees.cpp:
```diff
@@ -133,8 +133,88 @@ class TxConfirmStats
+
+int TxConfirmStats::CheckConsistent(size_t expected_bucket_count) const
+{
+    if (decay <= 0 || 1 <= decay) return 1;
```
Might be helpful to wrap these return values in an `struct/class enum` with variants that describe the consistency conditions. You could then downcast back to an `int` when doing arithmetic like in `CBlockPlicyEstimator::CheckConsistent`.

Then, in the test cases, check the calls to `CheckConsistent` match the expected variant, instead of just being non-zero.

ajtowns replied:
They don't have any meaning, they're just distinct to allow you to refer back to the line that found that inconsistency.

tryphe reviewed (commented):
Concept ACK involving estimation/stats/tests code changes
NACK on changing the default rate unless others really want it and there is good reasoning for it
light code review ACK c68f796c41b9b145c9c1b2c24aea505362ad1b63

I think it might be helpful to split this up into two PRs: one changing the minimum feerate variable(which doesn't necessarily do what it seems unless there is no demand for blockspace), and one involving estimation/stats/tests. There seems to be multiple interpretations and some confusion on what the default minimum should be (https://github.com/bitcoin/bitcoin/issues/14012#issuecomment-820363879), but I think the other changes are good changes.

prayank23 commented:
I had asked a related [question](https://bitcoin.stackexchange.com/questions/100861/why-cant-default-minimum-fee-rate-be-changed-to-0-1-0-2-sat-vbyte) in December 2020. Pieter Wuille had answered it with following things:

1. > Why can't we change default minimum fee rate to something below 1 sat/vByte?

> The primary reason for the minimum feerate (and related things like the discard feerate) is preventing the use of the P2P network as a cheap global broadcast system. If it's easy to produce transaction that will never confirm yet will relay across the network, it'd open the network to abuse. I don't know what the right level is to make sure that doesn't happen, but 1 sat/vbyte seems pretty low

2. > It was done in past: https://github.com/bitcoin/bitcoin/pull/3305 why not again?

> That was before the network and mempool logic had floating fees. Now the feerate is primarily set by market pressure, and lowering the minimum would at most have an impact during times when there is ~no demand for block space. I personally think the ecosystem should expect to function fine when there is continuous back pressure, so I'm not sure what is gained by reducing it.

3. > Do you think it will result in spamming? I don't think so and few other people as well: https://www.reddit.com/r/Bitcoin/comments/kjrbny/should_default_minimum_fee_rate_be_01_satvbyte_now/ggyym67

> I don't think spam is the problem. Lately the feerate has been significantly higher than 1 sat/vbyte anyway, and long term I hope that whatever the minimum is set to, it's effectively never hit. What matters is the discard feerate, which is not set by the market (it's the "additional" feerate you need to pay to replace a transaction in the mempool through RBF or mempool limiting, in order to make sure there is some cost to relaying the transaction that is now expected to not be mined anymore). That discard feerate is by default set to 1 sat/vbyte as well, which seems low.

After reading this answer and looking at few stats I am not sure if this is required. Also there is no way to test if lower fee rates will increase transactions and fee revenue for miners or decrease. Or if someone can find better charts and ways to test this before being used then we can re-evaluate. If we don't have lower default min fee rates, fee estimation for lower fee rates doesn't make sense so I have no opinion on this or weak NACK for now:

![image](https://user-images.githubusercontent.com/13405205/127573869-406ed8e3-6861-4f12-87b8-9b40974f5845.png)

ajtowns commented:
@prayank23 

> Still trying to understand what can go wrong if fee rate below 1 sat/vByte is allowed

This PR isn't changing that, so it's not entirely on-topic, suggest discussing elsewhere (eg on stackoverflow or in a PR that does change the fee rate or perhaps creating an issue on the topic if there isn't already one). There's two sorts of spam made more accessible by lowering the fee rate: spam that makes it into what would otherwise have been non-full blocks, which increases the resources needed for IBD; and spam that only ever goes through the p2p network and mempool, increasing the bandwidth cost of running a full node -- if you change the min fee to 0.1 sat/vbyte but keep the min incremental relay fee were left at 1 sat/vbyte, this might be pretty minimal though.

@rebroad 

> sorry for the noob-like question, but why is so much code needed to simply reduce the minfee...?

It's not changing the minfee, it's changing the range of values the fee estimation code supports. That needs a bunch of code mostly in order to support updating the old data set with a smaller range to a new data set with a larger range.

If you did the "one-line change" to update the minfee without updating the fee estimates, you'd always be told you need to pay at least 1sat/vbyte for your txs anyway; if you did it without the update code, fee estimates would only go below 1sat/vbyte for new installs or if you delete the `fee_estimates.dat` file.

@tryphe

> I think it might be helpful to split this up into two PRs: one changing the minimum feerate variable

This PR doesn't change the minimum feerate variable -- that's `DEFAULT_MIN_RELAY_TX_FEE` (validation.h) and `DEFAULT_INCREMENTAL_RELAY_FEE` (policy/policy.h) for mempool, and also `WALLET_INCREMENTAL_RELAY_FEE` (wallet/wallet.h) for the wallet.

ajtowns commented:
@MarcoFalke thanks for the ping; somehow I'd forgotten that it needed a rebase and followups

Rebased, responded to comments. If someone wants to take this over, they're welcome to -- I don't see this as a huge priority while a full 1sat/vb block would leave fees as less than 1% of the block subsidy. But better to have it open and potentially mergeable than end up forgotten and completely out of date.

prayank23 commented:
> This PR isn't changing that, so it's not entirely on-topic, suggest discussing elsewhere (eg on stackoverflow or in a PR that does change the fee rate or perhaps creating an issue on the topic if there isn't already one). There's two sorts of spam made more accessible by lowering the fee rate: spam that makes it into what would otherwise have been non-full blocks, which increases the resources needed for IBD; and spam that only ever goes through the p2p network and mempool, increasing the bandwidth cost of running a full node -- if you change the min fee to 0.1 sat/vbyte but keep the min incremental relay fee were left at 1 sat/vbyte, this might be pretty minimal though.

This PR is not needed if we don't have fee rates below 1 sat/vbyte.

prayank23 reviewed (changes requested):
Concept NACK

Fee rates below 1 sat/vbyte do not work right now and there are no plans or benefits for changing default minimum anytime soon. So fee estimation is not required for things that do not exist or relayed.

ajtowns commented:
> Fee rates below 1 sat/vbyte do not work right now and there are no plans or benefits for changing default minimum anytime soon. So fee estimation is not required for things that do not exist or relayed.

Fee rate below 1 sat/vbyte do work right now; they just require changing the default. My nodes where I have change the default have received transactions offering such low fee rates, and occassionally they are mined -- that is, while rare, they do exist and are relayed on current mainnet.

Having fee estimation work correctly with lower fee rates should be done *prior* to changing the defaults, not after, so that software attempting to work with the new defaults does not misbehave in unexpected ways.

prayank23 commented:
Such experiments do not require changes proposed in this PR which is a slippery slope to make things worse for fee market which already has issues.

I have tried experimenting with fee rate below 1 sat/vbyte: https://bitcoin.stackexchange.com/questions/99089/can-we-use-fee-rate-less-than-1-sat-vbyte/

Xekyo commented:
Seems reasonable, definitely agree that making everything else ready before changing the `DEFAULT_MIN_RELAY_TX_FEE` is the way to go. But if we're looking at a full block at minRelayTxFee bringing more than 1% of the fees, that's about a decade away, so I agree that it doesn't seem very urgent.

ajtowns commented:
Rebased, but is failing fuzz testing for reasons that aren't immediately obvious, so converted to draft

stevenroose commented:
Lately definitely looks like it is less relevant. Nothing says we can't go back to the situation from a year or two ago where txs with 1 sat/vB would easily confirm in a few hours.

achow101 commented:
Closing as up for grabs due to lack of activity.
